Source/Core/AudioCommon/AudioDevice.cpp

```diff
@@ -0,0 +1,212 @@
+{
+	devices.clear();
+
+#if defined _WIN32
+#define THROW_ON_ERROR(hr) if(FAILED(hr)) { throw hr; }
+
+	HRESULT hr;
+
+	IMMDeviceEnumerator* device_enumerator = NULL;
+	IMMDeviceCollection* device_collection = NULL;
+	u32 count = 0;
+
+	// Some awful COM code to get the number of audio endpoints.
+	try
+	{
+		hr = CoCreateInstance(__uuidof(MMDeviceEnumerator), NULL, CLSCTX_ALL, __uuidof(IMMDeviceEnumerator), (void**)&device_enumerator);
+		THROW_ON_ERROR(hr);
+
+		hr = device_enumerator->EnumAudioEndpoints(EDataFlow::eRender, DEVICE_STATE_ACTIVE, &device_collection);
+		THROW_ON_ERROR(hr);
+
+		hr = device_collection->GetCount(&count);
+		THROW_ON_ERROR(hr);
+	}
+	catch (HRESULT hresult)
+	{
+		ERROR_LOG(AUDIO, "Error while loading audio device list: %#X", hresult);
+	}
+
+	// Retrieve details for each audio endpoint.
+	for (u32 i = 0; i < count; i++)
+	{
+		AudioDevice device;
+		device.index = i;
+
+		LPWSTR id = NULL;
+		IPropertyStore* propStore = NULL;
+		IMMDevice* device_endpoint = NULL;
+
+		PROPVARIANT devPath;
+		PROPVARIANT devName;
+
+		// some more awful COM code to retrieve the necessary device information.
+		try
+		{
+			hr = device_collection->Item(i, &device_endpoint);
+			THROW_ON_ERROR(hr);
+
+			hr = device_endpoint->GetId(&id);
+			THROW_ON_ERROR(hr);
+			device.id = TStrToUTF8(id);
+			
+			hr = device_endpoint->OpenPropertyStore(STGM_READ, &propStore);
+			THROW_ON_ERROR(hr);
+			
+			PropVariantInit(&devPath);
+			// wtf? the necessary propkey for device path isn't defined?
+			// see https://msdn.microsoft.com/en-us/library/windows/desktop/hh405048(v=vs.85).aspx comment 
+			hr = propStore->GetValue(PROPERTYKEY{ { 0x9c119480, 0xddc2, 0x4954,{ 0xa1, 0x50, 0x5b, 0xd2, 0x40, 0xd4, 0x54, 0xad } }, 1 }, &devPath);
+			THROW_ON_ERROR(hr);
+			device.path = TStrToUTF8(devPath.pwszVal);
+			
+			PropVariantInit(&devName);
+			hr = propStore->GetValue(PKEY_Device_FriendlyName, &devName);
+			THROW_ON_ERROR(hr);
+			device.name = TStrToUTF8(devName.pwszVal);
+			
+			devices.push_back(device);
+		}
+		catch (HRESULT hresult)
+		{
+			ERROR_LOG(AUDIO, "Error while loading audio device: %#X", hresult);
+		}
+
+		CoTaskMemFree(id);
+		PropVariantClear(&devPath);
+		PropVariantClear(&devName);
+		if (propStore != NULL)
+			propStore->Release();
+		if (device_endpoint != NULL)
+			device_endpoint->Release();
+		
+	}
+
+	if (device_enumerator != NULL)
+		device_enumerator->Release();
+	if (device_collection != NULL)
+		device_collection->Release();
+
+#undef THROW_ON_ERROR
+#elif defined HAVE_ALSA && HAVE_ALSA
+
+	// get all sound devices from all cards for pcm
+	char **hints;
+	int err = snd_device_name_hint(-1, "pcm", (void***)&hints);
+	if (err != 0)
+	{
+		ERROR_LOG(AUDIO, "Error while loading audio device list: %#X", err);
+		return;
+	}
+
+	// iterate over all retrieved sound devices
+	char** n = hints;
+	snd_pcm_t* handle;
+	while (*n != NULL)
+	{
+
+		char *name = snd_device_name_get_hint(*n, "NAME");
```

Comment:
```cpp
std::unique_ptr<char, decltype(std::free)*> name { snd_device_name_get_hint(*n, "NAME"), std::free };
```

same with the others, then you don't need to care about calling free on it, it's done for you.